consumer.go

```diff
@@ -181,35 +189,50 @@ func (c *Consumer) ScanShard(
 
 			// loop records of page
 			for _, r := range resp.Records {
-				status := fn(r)
-
-				if !status.SkipCheckpoint {
-					lastSeqNum = *r.SequenceNumber
-
-					if err := c.checkpoint.Set(c.streamName, shardID, lastSeqNum); err != nil {
-						return err
-					}
-				}
-
-				if err := status.Error; err != nil {
+				lastSeqNum, isScanStopped, err = c.handleRecord(shardID, r, fn)
+				if err != nil {
 					return err
 				}
-
-				c.counter.Add("records", 1)
-
-				if status.StopScan {
+				if isScanStopped {
 					return nil
 				}
 			}
 
-			if resp.NextShardIterator == nil || shardIterator == resp.NextShardIterator {
+			if isShardClosed(resp.NextShardIterator, shardIterator) {
 				return nil
 			}
 			shardIterator = resp.NextShardIterator
 		}
 	}
 }
 
+func isShardClosed(nextShardIterator, currentShardIterator *string) bool {
```

Comment:
nice 👍

Reply:
More readable if we abstract the logic behind a function :)